### PR TITLE
Allow model instantiator relationship nulls

### PR DESCRIPTION
**Pull request by ha17**
Previously arrayToDataObject didn't allow a relationship to be null, even if the data object declared it nullable.

Note: _I originally branched from `main` thinking this repo was one that branched from `main` but afterwards pulled in `develop` into this branch, which brought in two commits._

### CODE REVIEW

**Review by timallen527 — APPROVED** [review]

File: src/Models/ModelInstantiator.php [hunk 1]
```diff
@@ -325,4 +329,28 @@ public function convertPropertyOnRetrieve(
 
         return $column->getConverter()->onRetrieve($property);
     }
+
+    /**
+     * @param mixed $property
+     * @param array $property_reflection
+     * @return mixed
+     */
+    private function getReflectionProperty(mixed $property, array $property_reflection): mixed
```
[the remark this answers is not in the document]
Comment (pcrist) [reply src/Models/ModelInstantiator.php:338]: Looks like the return type may always be an array.

File: src/Models/Test/ModelInstantiatorTest.php [hunk 2]
```diff
@@ -156,6 +156,55 @@ public function testArrayToDataObjectBadNestedObject(): void
...
+            $array,
+            ModelInstantiatorTestObject::class,
+        );
+
+        $this->assertEquals(
+            $data_object,
+            $actual
+        );
+    }
+
```
Comment (pcrist) [inline src/Models/Test/ModelInstantiatorTest.php:207]: Do we need a test for a non-nullable relationship that is defined in the incoming array as null?

File: src/Models/ModelInstantiator.php [hunk 3]
```diff
@@ -83,6 +82,12 @@ public function arrayToDataObject(
                 $instance->{$property} = $instances;
             } else {
                 // ManyToOne: Map to a single nested Data Object.
+
+                if (is_null($arr[$property]) && $this->propertyAllowsNull($property, $property_reflection)) {
```
Comment (pcrist) [inline src/Models/ModelInstantiator.php:86]: It may make sense to use `$enforcer->propIsNull()` here: https://github.com/PDGA/php-data-objects/blob/9f8eb4ee428df91c9c516a6d5d67492be6125a29/src/Enforcers/ValidationEnforcer.php#L147

That would add another call to `propIsUndefined`; not sure if it's worth the tradeoff.

**Review by pcrist — APPROVED** [review]
Looks good; potential for some unit tests and more specific type-hinting.

File: src/Models/ModelInstantiator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (ha17) [reply src/Models/ModelInstantiator.php:338]: > Looks like the return type may always be an array.

This array coming in is an array of ReflectionProperty objects:

```
Array
(
    [0] => ReflectionProperty Object
        (
            [name] => pdgaNumber
            [class] => PDGA\DataObjects\Models\Test\ModelInstantiatorTestObject
        )

    [1] => ReflectionProperty Object
        (
            [name] => firstName
            [class] => PDGA\DataObjects\Models\Test\ModelInstantiatorTestObject
        )
```

It's a little confusing because the `array_column` function can take an array of arrays or array of objects. In this case it's objects and it's pulling out the `name` property and searching it.

In this case, I changed `mixed` to `ReflectionProperty` and the tests pass still.

File: src/Models/Test/ModelInstantiatorTest.php (hunk 2 above)
Comment (ha17) [reply src/Models/Test/ModelInstantiatorTest.php:207]: Good call; I've added this test

File: src/Models/ModelInstantiator.php [hunk 4]
```diff
@@ -331,23 +332,34 @@ public function convertPropertyOnRetrieve(
...
+     * @param string $property ReflectionProperties->getName() always returns
+     *                         a string which is the key of the array that
+     *                         handles this property.
      * @param array $property_reflection
-     * @return mixed
+     * @return \ReflectionProperty
+     * @throws \TypeError
      */
-    private function getReflectionProperty(mixed $property, array $property_reflection): mixed
+    public function getReflectionProperty(string $property, array $property_reflection): \ReflectionProperty
```
Comment (timallen527) [inline src/Models/ModelInstantiator.php:343]: We should add the use statement for `ReflectionProperty` instead of using the `\`.

File: src/Models/ModelInstantiator.php [hunk 5]
```diff
@@ -325,4 +330,39 @@ public function convertPropertyOnRetrieve(
...
+     * @param array $property_reflection
+     * @return \ReflectionProperty
+     * @throws \TypeError
+     */
+    public function getReflectionProperty(string $property, array $property_reflection): \ReflectionProperty
+    {
+        $reflection_index = array_search($property, array_column($property_reflection, 'name'), true);
+
+        if (false === $reflection_index || !isset($property_reflection[$reflection_index])) {
+            throw new \TypeError();
```
Comment (timallen527) [inline src/Models/ModelInstantiator.php:348]: Is `TypeError` the appropriate Exception type here? Maybe `OutOfBoundsException` with an appropriate message is more correct.

https://www.php.net/manual/en/class.outofboundsexception.php

Also I think we should add the use statement and get rid of the `\`

File: src/Models/Test/ModelInstantiatorTest.php [hunk 6]
```diff
@@ -593,6 +669,68 @@ public function testDatabaseModelToDataObjectNestedObjectWithRelationsGetterExce
...
+        $property_reflection = $reflection_class->getProperties();
+        $property = 'nullableFakeHasOneRelation'; // this is defined as being nullable
+
+        // Search to make sure the property exists in the data object or we'll get a false
+        // positive for this test.
+        $found = array_search($property, array_column($property_reflection, 'name'));
+        $result = $this->model_instantiator->propertyAllowsNull($property, $property_reflection);
+
+        $this->assertTrue($result);
+        $this->assertGreaterThan(0, $found);
```
Comment (timallen527) [inline src/Models/Test/ModelInstantiatorTest.php:689]: If the `$found` is necessary for this test to be accurate, it might make sense to move this assertion to right below  where it is defined. Then maybe add a failure message to the assertion that explains why the setup is incorrect instead of just having the comment. This could be applied to the test below as well.

File: src/Models/Test/ModelInstantiatorTest.php [hunk 7]
```diff
@@ -593,6 +669,68 @@ public function testDatabaseModelToDataObjectNestedObjectWithRelationsGetterExce
...
+
+    public function testGetReflectionPropertyCorrectlyReturnsReflectionPropertyObject()
+    {
+        $reflection_class = new ReflectionClass(ModelInstantiatorTestObject::class);
+        $property_reflection = $reflection_class->getProperties();
+        $property = 'fakeHasOneRelation'; // this property should exist on the test data object
+
+        $result = $this->model_instantiator->getReflectionProperty($property, $property_reflection);
+
+        $this->assertInstanceOf(\ReflectionProperty::class, $result);
```
Comment (timallen527) [inline src/Models/Test/ModelInstantiatorTest.php:720]: The `\`s here and in the next test can be replaced with use statements.

**Review by timallen527 — APPROVED** [review]
Thanks for the updates. The only other thing that might be helpful is for that `OutOfBoundsException` to have a message when thrown.

**Review by pcrist — APPROVED** [review]
